File: avatar.go
// avatar.go
package main

import (
	"flag"
	"fmt"
	"hash/fnv"
	"image"
	"image/color"
	"image/draw"
	"image/jpeg"
	"image/png"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/fogleman/gg"
)

func hashString(s string) uint32 {
	h := fnv.New32a()
	h.Write([]byte(s))
	return h.Sum32()
}

func hashToColor(h uint32) color.RGBA {
	return color.RGBA{
		R: uint8((h >> 16) & 0xFF),
		G: uint8((h >> 8) & 0xFF),
		B: uint8(h & 0xFF),
		A: 255,
	}
}

func parseHexColor(c string) color.RGBA {
	if len(c) == 7 && c[0] == '#' {
		r, _ := strconv.ParseInt(c[1:3], 16, 8)
		g, _ := strconv.ParseInt(c[3:5], 16, 8)
		b, _ := strconv.ParseInt(c[5:7], 16, 8)
		return color.RGBA{uint8(r), uint8(g), uint8(b), 255}
	}
	return color.RGBA{0, 0, 0, 255}
}

func generateAvatar(name, output string, size int, fmt string, mainColor, bgColor string, addText bool) {
	hash := hashString(name)
	mainCol := hashToColor(hash)
	bgCol := hashToColor(hash + 12345)

	if mainColor != "" {
		mainCol = parseHexColor(mainColor)
	}
	if bgColor != "" {
		bgCol = parseHexColor(bgColor)
	}

	dc := gg.NewContext(size, size)
	dc.SetColor(bgCol)
	dc.Clear()

	// Рисуем круг
	margin := float64(size) / 20.0
	dc.SetColor(mainCol)
	dc.DrawCircle(float64(size)/2, float64(size)/2, float64(size)/2-margin)
	dc.Fill()

	if addText {
		// Инициалы
		initials := ""
		parts := strings.Split(name, " ")
		for _, p := range parts {
			if len(p) > 0 {
				initials += string(p[0])
			}
		}
		if initials == "" {
			initials = string(name[0])
		}
		initials = strings.ToUpper(initials)
		fontSize := float64(size) / 3.0
		dc.SetColor(color.RGBA{255, 255, 255, 255})
		if err := dc.LoadFontFace("/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf", fontSize); err != nil {
			// fallback
			dc.SetRGB(1, 1, 1)
		}
		dc.DrawStringAnchored(initials, float64(size)/2, float64(size)/2, 0.5, 0.5)
	}

	// Сохранение
	var err error
	switch fmt {
	case "png":
		err = dc.SavePNG(output)
	case "jpg", "jpeg":
		err = dc.SaveJPG(output, 90)
	default:
		err = dc.SavePNG(output)
	}
	if err == nil {
		fmt.Printf("✅ Аватарка сохранена в %s\n", output)
	} else {
		fmt.Printf("❌ Ошибка: %v\n", err)
	}
}

func main() {
	var (
		name      string
		output    string
		size      int
		format    string
		mainColor string
		bgColor   string
		addText   bool
		random    bool
	)
	flag.StringVar(&name, "name", "", "Имя")
	flag.StringVar(&output, "o", "avatar.png", "Выходной файл")
	flag.IntVar(&size, "s", 256, "Размер")
	flag.StringVar(&format, "f", "png", "Формат (png,jpg,webp)")
	flag.StringVar(&mainColor, "c", "", "Основной цвет (HEX)")
	flag.StringVar(&bgColor, "b", "", "Цвет фона (HEX)")
	flag.BoolVar(&addText, "t", false, "Добавить текст")
	flag.BoolVar(&random, "r", false, "Случайная генерация")
	flag.Parse()

	if flag.NArg() > 0 {
		name = flag.Arg(0)
	}

	if random {
		rand.Seed(time.Now().UnixNano())
		name = fmt.Sprintf("User%d", rand.Intn(10000))
	}
	if name == "" {
		fmt.Println("Укажите имя или используйте --random")
		os.Exit(1)
	}

	generateAvatar(name, output, size, format, mainColor, bgColor, addText)
}

File: avatar.py
# avatar.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import hashlib
import argparse
import random
from PIL import Image, ImageDraw, ImageFont, ImageColor
import math

def hash_to_color(hash_val):
    """Преобразует хеш в цвет (HEX)."""
    r = (hash_val >> 16) & 0xFF
    g = (hash_val >> 8) & 0xFF
    b = hash_val & 0xFF
    return (r, g, b)

def generate_avatar(name, size=256, output="avatar.png", fmt="png", main_color=None, bg_color=None, add_text=False):
    """Генерирует аватарку на основе имени."""
    if not main_color:
        hash_val = int(hashlib.md5(name.encode()).hexdigest()[:8], 16)
        main_color = hash_to_color(hash_val)
    else:
        main_color = ImageColor.getrgb(main_color)

    if not bg_color:
        bg_hash = int(hashlib.md5((name + "bg").encode()).hexdigest()[:8], 16)
        bg_color = hash_to_color(bg_hash)
    else:
        bg_color = ImageColor.getrgb(bg_color)

    # Создаём изображение
    img = Image.new('RGB', (size, size), bg_color)
    draw = ImageDraw.Draw(img)

    # Рисуем круг
    margin = size // 20
    draw.ellipse((margin, margin, size - margin, size - margin), fill=main_color)

    # Добавляем инициалы
    if add_text:
        initials = ''.join(word[0].upper() for word in name.split() if word)
        if not initials:
            initials = name[0].upper()
        try:
            font_size = size // 3
            font = ImageFont.truetype("/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf", font_size)
        except:
            font = ImageFont.load_default()
        bbox = draw.textbbox((0, 0), initials, font=font)
        text_w = bbox[2] - bbox[0]
        text_h = bbox[3] - bbox[1]
        x = (size - text_w) // 2
        y = (size - text_h) // 2
        draw.text((x, y), initials, fill=(255, 255, 255), font=font)

    # Сохраняем
    img.save(output, format=fmt.upper())

def batch_generate(names, **kwargs):
    """Генерирует аватарки для списка имён."""
    for name in names:
        output = f"avatar_{name}.png"
        generate_avatar(name, **kwargs, output=output)
        print(f"✅ {name} -> {output}")

def main():
    parser = argparse.ArgumentParser(description="Avatar Generator")
    parser.add_argument('name', nargs='?', default=None, help='Имя для генерации')
    parser.add_argument('-o', '--output', default='avatar.png', help='Выходной файл')
    parser.add_argument('-s', '--size', type=int, default=256, help='Размер в пикселях')
    parser.add_argument('-f', '--format', choices=['png', 'jpg', 'webp'], default='png', help='Формат')
    parser.add_argument('-c', '--color', help='Основной цвет (HEX)')
    parser.add_argument('-b', '--bg', help='Цвет фона (HEX)')
    parser.add_argument('-t', '--text', action='store_true', help='Добавить текст (инициалы)')
    parser.add_argument('-r', '--random', action='store_true', help='Случайная генерация')
    parser.add_argument('--batch', help='Пакетная генерация (имена через запятую)')
    parser.add_argument('--batch-file', help='Файл со списком имён (по одному на строку)')
    args = parser.parse_args()

    if args.random:
        names = [f"User{random.randint(1000, 9999)}"]
    elif args.batch:
        names = [n.strip() for n in args.batch.split(',') if n.strip()]
    elif args.batch_file:
        with open(args.batch_file, 'r') as f:
            names = [line.strip() for line in f if line.strip()]
    elif args.name:
        names = [args.name]
    else:
        print("Укажите имя или используйте --random", file=sys.stderr)
        sys.exit(1)

    kwargs = {
        'size': args.size,
        'fmt': args.format,
        'main_color': args.color,
        'bg_color': args.bg,
        'add_text': args.text,
    }

    if len(names) == 1:
        output = args.output or f"avatar_{names[0]}.png"
        generate_avatar(names[0], output=output, **kwargs)
        print(f"✅ Аватарка сохранена в {output}")
    else:
        batch_generate(names, **kwargs)

if __name__ == '__main__':
    main()
